src/library/clementine/clementinedbconnection.cpp

```diff
@@ -0,0 +1,183 @@
+                          .arg(playlistId);
+
+    query.prepare(queryString);
+    if (query.exec()) {
+        while (query.next()) {
+            QString type = query.value(16).toString();
+            if (type != "File")
+                continue;
+
+            //Search for track in mixxx lib to provide bpm information
+            QString location = QUrl::fromEncoded(
+                    query.value(2).toByteArray(), QUrl::StrictMode)
+                                       .toLocalFile();
+            bool track_already_in_library = false;
+            TrackPointer pTrack = m_pTrackCollectionManager->getOrAddTrack(
+                    TrackRef::fromFileInfo(location),
+                    &track_already_in_library);
+
+            entry.artist = query.value(4).toString();
+            entry.title = query.value(1).toString();
+            entry.trackId = query.value(0).toInt();
+            long long duration = query.value(3).toLongLong();
+            entry.year = query.value(5).toInt();
+            entry.album = query.value(6).toString();
+            entry.albumartist = query.value(17).toString();
+            entry.uri = QUrl::fromEncoded(query.value(2).toByteArray(), QUrl::StrictMode);
+            entry.rating = query.value(7).toInt();
+            entry.genre = query.value(8).toString();
+            entry.grouping = query.value(15).toString();
+            entry.tracknumber = query.value(9).toInt();
+            entry.bpm = query.value(10).toInt();
+            entry.bitrate = query.value(11).toInt();
+            entry.comment = query.value(12).toString();
+            entry.playcount = query.value(13).toInt();
+            entry.composer = query.value(14).toString();
+
+            if (entry.artist == "" && entry.title == "") {
+                entry.artist =
+                        "Unknown"; // may confuse with real "Unknown" artists from whitelabels?
+                entry.title = location.split(QDir::separator()).last();
+            }
+            if (entry.bpm == -1) {
+                entry.bpm = 0;
+            }
+            if (duration == -1) {
+                entry.duration = 0;
+            } else {
+                entry.duration = int(duration / 1000000000);
+            }
+
+            //If found in mixxx lib overwrite information
+            if (track_already_in_library) {
+                entry.bpm = pTrack->getBpm();
+                entry.duration = pTrack->getDurationInt();
+            }
+
+            list.append(entry);
+        }
+    } else {
+        LOG_FAILED_QUERY(query);
+    }
+
+    qDebug() << "ClementineDbConnection::getPlaylistEntries(), took"
+             << time.elapsed().debugMillisWithUnit();
+
+    return list;
+}
+
+// static
+QString ClementineDbConnection::getDatabaseFile() {
+    QString dbfile;
+
+    QSettings ini(QSettings::IniFormat, QSettings::UserScope, "Clementine", "Clementine");
+    dbfile = QFileInfo(ini.fileName()).absolutePath();
+    dbfile += "/clementine.db";
```

Comment:
Let's not hardcode dir separators. Clemetine also runs on Windows.

```suggestion
    QDir configDir QFileInfo(ini.fileName()).dir();
    dbfile = configDir.absoluteFilePath("clementine.db");
```

Reply:
will keep this for now due to comment below